Secure Expense Dashboard with Express and Okta Blog (#1584)
oktadeveloper/okta-blog: _source/_posts/2025-07-28-express-oauth-pkce.md

Initialize a new Node.js project:

```sh
npm init -y
```

Install the required packages:

```sh
npm install express@5.1 passport@0.7 openid-client@6.6 express-session@1.18 ejs@3.1 express-ejs-layouts@2.5 dotenv
```

Now, install the development dependencies:

```sh
npm install --save-dev nodemon
``` 

**What do these dependencies do?** 

These installed packages become your Express project's dependencies.

* **`express`**: Handles routing and HTTP middleware for your web app

* **`passport`**: Sets up and maintains server-side sessions

* **`openid-client`**: Node.js OIDC library with PKCE support; handles the OAuth handshake and token exchange.

* **`express-session`**: Manages user sessions on the server

* **`dotenv`**: Loads environment variables from a `.env` file

* **`ejs`**: Enables dynamic HTML rendering using embedded JavaScript templates

* **`express-ejs-layouts`**: Adds layout support to EJS, helping manage common layout structures across views

## Configure environment variables for OIDC authentication

Create a `.env` file in the root directory with placeholders for your Okta configuration.

```

> OKTA_ISSUER=https://{yourOktaDomain}/oauth2/default 

Reviewer: ```suggestion
OKTA_ISSUER=https://{yourOktaDomain}
```

Are you using the custom authorization server? if not, then this is the correct URL

OKTA_CLIENT_ID={yourClientId}
OKTA_CLIENT_SECRET={clientSecret}
APP_BASE_URL=http://localhost:3000
POST_LOGOUT_URL=http://localhost:3000
```

In the next step, you'll get these values from your Okta Admin Console.

## Create the Okta OIDC web application

> {% include setup/integrator.md type="web" loginRedirectUri="http://localhost:3000/authorization-code/callback" logoutRedirectUri="http://localhost:3000" %}

Reviewer: Since you don't have the custom AS param here, you aren't using a custom AS. 

Note, please step through the post step by step (including setting up a new okta integrator account) to ensure your instructions are valid for it. As you have it right now, this post wouldn't work.

    },
    {
      name: "Grace Li",
      item: "Compliance Training",
      amount: 500,
    },
    {
      name: "Alice Johnson",
      item: "Team Offsite",
      amount: 950,
    },
  ],
};

export function getModifiedTeam(team) {
  if (!team?.trim()) return [];

  const toPascalCase = (str) =>
    str
      .trim()
      .split(/\s+/)
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
      .join(' ');

  const toKebabCase = (str) => str.trim().toLowerCase().split(' ').join('-');

  if (team === 'admin') {
    return ALL_TEAMS_NAME.map((element) => ({
      id: toKebabCase(element),
      label: toPascalCase(element),
    }));
  }

  return [
    {
      id: toKebabCase(team),
      label: toPascalCase(team),
    },
  ];
}
```

The file also defines `getModifiedTeam`, a helper that converts a team name into an array of objects. Each object has an id and a label. If the team is admin, the function returns an object for every entry in `ALL_TEAMS_NAME`; otherwise, it returns a single object for the specified team. Later in the project, the app calls this function to transform the user's team information.

### Create a file to handle authentication

> Create an `auth.js` file for this step. This file uses the [openid-client](https://www.passportjs.org/packages/openid-client) library to handle the OIDC flow: it logs users in, exchanges the authorization code for tokens, and logs them out. It also defines a middleware that guards protected routes.

Reviewer: ```suggestion
Create an `auth.js` file for this step. This file uses the [openid-client](https://www.npmjs.com/package/openid-client) library to handle the OIDC flow: it logs users in, exchanges the authorization code for tokens, and logs them out. It also defines a middleware that guards protected routes.
```

Recommend either the npm or GH URL instead of the package through Passport (as we're not using it as a Passport strategy)

      name,
      email,
      teams,
      idToken: tokenSet.id_token,
    };

    delete req.session.pkce;

    req.logIn(userProfile, (err) => {
      if (err) {
        return next(err);
      }

      return res.redirect("/dashboard");
    });
  } catch (error) {
    console.error("Authentication error:", error.message);
    return res.status(500).send(`Authentication failed: ${error.message}`);
  }
}

export async function logout(req, res) {
  try {
    const openIdClientConfig = await getClientConfig();

    const id_token_hint = req.user?.idToken;

    const logoutUrl = client.buildEndSessionUrl(openIdClientConfig, {
      id_token_hint,
      post_logout_redirect_uri: process.env.POST_LOGOUT_URL,
    });

    req.logout((err) => {
      if (err) return next(err);

      req.session.destroy((err) => {
        if (err) return next(err);
        res.redirect(logoutUrl);
      });
    });
  } catch (error) {
    res.status(500).send('Something went wrong during logout.');
  }
}

export function ensureAuthenticated(req, res, next) {
  if (req.isAuthenticated()) {
    return next();
  }
  res.redirect("/login");
}
```

This file includes the following functions:

> 1. **`getClientConfig`** - Discovers the authorization server's metadata and returns the associated metadata object.

Reviewer: ```suggestion
1. **`getClientConfig`** - Retrieves the authorization server's metadata using the authorization server's discovery endpoint.
```

> 2. **`login`** - This function starts the Authorization Code + PKCE flow. It generates a **code_verifier**, derives a **code_challenge**, and creates a state value. It stores the **code_verifier** and **state** in the session, builds the authorize URL, and redirects the user to Okta to complete the login flow.

Reviewer: ```suggestion
2. **`login`** - This function starts the Authorization Code + PKCE flow. It generates the required values to enable PKCE: the **code_verifier** and **code_challenge**. These values, along with the `state` value protect the user sign in process from attack vectors. PKCE protects against auth code interception attacks and the state parameter protects against Cross-site Request Forgery (CSRF). The openid-client builds the user sign in URL with these values and redirects the user to Okta to complete the authentication challenge.
```

Explaining a bit more about why each value instead of rehashing the code. Might want to consider linking to PKCE resource and CSRF too?

Reviewer: Not Grammarly checked

3. **`getCallbackUrlWithParams`** - Reconstructs the complete callback URL, including protocol, host, path, and query.

> 4. **`authCallback`** - Retrieves the **code_verifier** and **state** from the session and use it to obtain tokens. Map the user to a team and store the profile details and ID token in the session. If everything succeeds, redirect the user to the dashboard.

Reviewer: ```suggestion
4. **`authCallback`** - This function runs when the user redirects back to the app after the authentication challenge succeeds. At this point, the redirect URL back into the application includes the auth code. The OIDC client verifies the redirect URL by checking that the **state** value matches the parameter it sent previously. Once verified, the OIDC uses the auth code for the token exchange by adding the **code_verifier** to the token request. The authorization server validates the auth code and the code_verifier value to ensure the request comes from the client making the original authentication request, mitigating attacks using stolen auth codes.

Once we get back valid tokens, we handle the app's business logic, such as mapping the user to a team and storing the profile details and ID token in the session. If everything succeeds, it redirects the user to the dashboard.
```

      name,
      email,
      teams,
      idToken: tokenSet.id_token,
    };

    delete req.session.pkce;

    req.logIn(userProfile, (err) => {
      if (err) {
        return next(err);
      }

      return res.redirect("/dashboard");
    });
  } catch (error) {
    console.error("Authentication error:", error.message);
    return res.status(500).send(`Authentication failed: ${error.message}`);
  }
}

export async function logout(req, res) {
  try {
    const openIdClientConfig = await getClientConfig();

    const id_token_hint = req.user?.idToken;

    const logoutUrl = client.buildEndSessionUrl(openIdClientConfig, {
      id_token_hint,
      post_logout_redirect_uri: process.env.POST_LOGOUT_URL,
    });

    req.logout((err) => {
      if (err) return next(err);

      req.session.destroy((err) => {
        if (err) return next(err);
        res.redirect(logoutUrl);
      });
    });
  } catch (error) {
    res.status(500).send('Something went wrong during logout.');
  }
}

export function ensureAuthenticated(req, res, next) {
  if (req.isAuthenticated()) {
    return next();
  }
  res.redirect("/login");
}
```

This file includes the following functions:

> 1. **`getClientConfig`** - Retrieves the authorization server's metadata using authorization server's discovery endpoint.

Reviewer: ```suggestion
1. **`getClientConfig`** - Retrieves the authorization server's metadata using the discovery endpoint.
```

2. **`login`** - This function starts the Authorization Code + PKCE flow. It generates the required values to enable PKCE: the **code_verifier** and **code_challenge**. These values, along with the `state` value protect the user sign in process from attack vectors. [PKCE](https://datatracker.ietf.org/doc/html/rfc7636) protects against auth code interception attacks, and the state parameter protects against [Cross-Site Request Forgery (CSRF)](https://owasp.org/www-community/attacks/csrf). The openid-client builds the user sign in URL with these values and redirects the user to Okta to complete the authentication challenge.
3. **`getCallbackUrlWithParams`** - Reconstructs the complete callback URL, including protocol, host, path, and query.

> 4. **`authCallback`** - This function runs when the user redirects back to the app after the authentication challenge succeeds. At this point, the redirect URL back into the application includes the auth code. The OIDC client verifies the auth code by checking that the **state** value matches the parameter in the first redirect. Once verified, the OIDC uses the auth code for the token exchange by adding the **code_verifier** to the token request. 

Reviewer: ```suggestion
4. **`authCallback`** - This function runs when the user redirects back to the app after the authentication challenge succeeds. At this point, the redirect URL back into the application includes the auth code. The OIDC client verifies the auth code by checking that the **state** value matches the parameter in the first redirect. Once verified, the openid-client library uses the auth code for the token exchange by adding the **code_verifier** to the token request. The authorization server validates the auth code and the `code_verifier` value to ensure the request comes from the client making the original authentication request, mitigating attacks using stolen auth codes.
```
